File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/barrel/AlchemyDisplay.java
package wraith.fabricaeexnihilo.compatibility.rei.barrel;

import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.display.Display;
import me.shedaniel.rei.api.common.entry.EntryIngredient;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import net.minecraft.item.SpawnEggItem;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;
import wraith.fabricaeexnihilo.recipe.barrel.AlchemyRecipe;

import java.util.List;
import java.util.Optional;

public class AlchemyDisplay implements Display {
    public final EntryIngredient catalyst;
    public final EntryIngredient reactant;
    public final EntryIngredient byproduct;
    public final EntryIngredient result;
    public final EntryIngredient entity;
    private final Identifier id;

    public AlchemyDisplay(AlchemyRecipe recipe) {
        this.catalyst = EntryIngredients.ofIngredient(recipe.getCatalyst());
        this.reactant = recipe.getReactant().asReiIngredient();
        this.result = recipe.getResult().getReiResult();
        this.byproduct = EntryIngredients.of(recipe.getByproduct().stack());
        this.entity = recipe.getToSpawn().isEmpty() ? EntryIngredient.empty() : EntryIngredients.of(SpawnEggItem.forEntity(recipe.getToSpawn().getType()));
        this.id = recipe.getId();
    }

    @Override
    public CategoryIdentifier<?> getCategoryIdentifier() {
        return PluginEntry.ALCHEMY;
    }

    @Override
    public List<EntryIngredient> getInputEntries() {
        return List.of(catalyst, reactant);
    }

    @Override
    public List<EntryIngredient> getOutputEntries() {
        return List.of(result, byproduct, entity);
    }

    @Override
    public Optional<Identifier> getDisplayLocation() {
        return Optional.of(id);
    }
}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/barrel/AlchemyDisplay.kt
package exnihilofabrico.compatibility.rei.barrel

import exnihilofabrico.api.recipes.barrel.AlchemyRecipe
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.modules.ModBlocks
import exnihilofabrico.modules.barrels.modes.FluidMode
import exnihilofabrico.modules.barrels.modes.ItemMode
import exnihilofabrico.util.asREIEntry
import me.shedaniel.rei.api.EntryStack
import me.shedaniel.rei.api.RecipeDisplay
import net.minecraft.item.SpawnEggItem

class AlchemyDisplay(val recipe: AlchemyRecipe): RecipeDisplay {
    override fun getRecipeCategory() = PluginEntry.ALCHEMY

    override fun getOutputEntries() = listOf(
        recipe.product.let { mode ->
            when(mode){
                is ItemMode -> mode.stack.asREIEntry()
                is FluidMode -> mode.fluid.rawFluid?.bucketItem?.asREIEntry() ?: EntryStack.empty()
                else -> EntryStack.empty()
            }
        },
        recipe.byproduct.stack.asREIEntry(),
        if(!recipe.toSpawn.isEmpty())
            SpawnEggItem.forEntity(recipe.toSpawn.type)?.asREIEntry()
        else
            EntryStack.empty()
    )
    override fun getInputEntries() =
        listOf(recipe.reactant.asREIEntries(), recipe.catalyst.asREIEntries(), ModBlocks.BARRELS.values.map { it.asREIEntry() })

}